### PR TITLE
Auto-refresh expired tokens

### PR DESCRIPTION
**Pull request by ajacksified**
Uses the refresh token to be able to keep a user logged in indefinitely.

:eyeglasses: @danehansen @curioussavage

### CODE REVIEW

File: src/server.jsx [hunk 1]
```diff
@@ -194,6 +196,29 @@ class Server {
...
+      if (!expires) {
+        yield next;
+        return;
+      }
+
+      expires = new Date(expires);
+
+      if (now > expires) {
+        var rToken = this.cookies.get('refreshToken');
+        var token = yield app.refreshToken(this, rToken);
```
Comment (ajacksified) [inline src/server.jsx]: Needs try/catch to kill the session if the refresh token request fails.

Comment (danehansen) [inline src/server.jsx]: what is "function * (next)"? not familiar with that. otherwise looks fine.

File: assets/js/client.es6.js [hunk 2]
```diff
@@ -47,13 +47,34 @@ function setTitle(props={}) {
   }
 }
 
+function refreshToken (app, tries=0) {
```
Comment (danehansen) [inline assets/js/client.es6.js]: oh god i didn't realize we could do default values in es6 thats rad

**Comment by curioussavage** [comment]
New syntax is a still a little foreign to me but It just took an extra read through.

So the token has to be refreshed every hour?

**Comment by ajacksified** [comment]
Correct, it expires after an hour, so we have to re-get the token before it expires.

Also :eyeglasses: @JordanMilne just in case I missed something.

File: src/oauth.es6.js [hunk 3]
```diff
@@ -7,26 +7,30 @@ import uuid from 'uuid';
...
-    maxAge: 1000 * 3000,
-  };
+  function setTokenCookie(ctx, token) {
+    ctx.cookies.set('token', token.token.access_token, longCookieOptions);
+    ctx.cookies.set('tokenExpires', token.token.expires_at.toString(), longCookieOptions);
 
-  if (app.getConfig('cookieDomain')) {
-    cookieOptions.domain = app.getConfig('cookieDomain');
+    if (token.token.refresh_token) {
+      ctx.cookies.set('refreshToken', token.token.refresh_token, longCookieOptions);
```
Comment (JordanMilne) [inline src/oauth.es6.js]: Can `refreshToken` be made `HttpOnly` if it's only read / written by `/oauth/refresh`? That'd allow us to minimize the impact of XSSes.

**Comment by JordanMilne** [comment]
:fish: Looks good, just one question

File: src/oauth.es6.js (hunk 3 above)
Comment (ajacksified) [inline src/oauth.es6.js]: Yup! longCookieOptions has `httpOnly` set to true.

Comment (JordanMilne) [inline src/oauth.es6.js]: doh, missed that!

**Pull request merged**